src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -1354,13 +1357,82 @@ public ValidateCountAttribute(int minLength, int maxLength)
     }
 
     /// <summary>
+    /// Base class for all valid values generators with cache support.
+    /// </summary>
+    public abstract class CachedValidValuesGeneratorBase : IValidateSetValuesGenerator
+    {
+        // Cached valid values.
+        private string[] _validValues;
+
+        /// <summary>
+        /// Sets a time interval in seconds to reset the '_validValues' dynamic valid values cache.
+        /// By default 'ValidValuesCacheExpiration = 0' and no valid values is cached (is re-evaluated every time it is used).
+        /// </summary>
+        public int ValidValuesCacheExpiration { 
```

Comment:
The property is not available for users from outside the class because we create an instance of the class by type with _default_ constructor only. We need way to set `_validValuesCacheExpiration`. See a sample from our tests below. So we cannot remove the property but we can make it as `protected`. 

```powershell
            class GenValuesWithExpiration : System.Management.Automation.CachedValidValuesGeneratorBase {
                GenValuesWithExpiration() : base() {
                    $this.ValidValuesCacheExpiration = 2
                }

                Static [bool] $temp = $true;

                [String[]] GenerateValidValues() {

                    if ([GenValuesWithExpiration]::temp) {
                        [GenValuesWithExpiration]::temp = $false
                        return [string[]]("Test1","TestString1","Test2")
                    } else {
                        [GenValuesWithExpiration]::temp = $true
                        return [string[]]("Test1","TestString2","Test2")
                    }
                    
                }
            }
```